### DISCUSSION TITLE
Fermata not showing up in launcher (Android 8.0)

### DISCUSSION THREAD

**Discussion by roger2k13324** [AndreyPavlenko/Fermata#392] [General]
Hello there!

First of all, I would like to thank you for creating such a great application. I am a software developer myself and know how much effort goes into things like Fermata.

I have been using Android Auto for years but somehow missed the fact that there are already workarounds making it possible to play videos on the head unit. I have and old Samsung Galaxy S7 with a broken but functional screen lying around and thought it would be the perfect device for that use. However, I could not manage to get Fermata to show up in the launcher when the phone is connected to the car head unit. With the desktop head unit emulator, the Fermata apps are visible.

Please note that I have not installed Fermata Control as I don't care about the steering wheel buttons.

The phone model is SM-G930F (Samsung Galaxy S7) with Android 8.0.0. Device is rooted with Magisk 27.0.

What I have tried:
	- AA Phenotype Patcher 0.85 (no luck)
	- AA AIO Tweaker 5.2.2 (no luck)
	- KingInstaller 1.4 (no luck)
	- Android Auto downgrade to different earlier versions (on connection to the HU it forces update to the current version)
	- Google Play Services downgrade to different earlier versions (incompatibility problems with current Android Auto version, see above)
	- Xposed Version 90-beta3 (latest compatible with Android 8.0) and Fermata module (no luck, NoSuchMethodException when hooking getInstallerPackageName in projection.gearhead)
	- LSPosed 1.0.2 (latest compatible with Android 8.0) + AAXLU does not work because LSPosed prompts me to install its own Magisk module (even though both Riru and LSPosed are already installed)

Furthermore, even with the DHU, only Fermata Mirror FS is actually working, both Fermata and Fermata Mirror produce a black surface and a hang of Android Auto.

When I connect the phone to the car HU after being connected to DHU, Fermata Mirror FS starts automatically (navigation app) and works fine, but if the phone gets disconnected and reconnected, it does not start anymore. I have to connect it to DHU again to make it work.

Do you have any idea what I could try considering the device is rooted? Is there any way to start an AA application from a root shell or by an intent?

Thank you!

**Comment by AndreyPavlenko** [comment]
Seems the method argument type is missing here - https://github.com/AndreyPavlenko/Fermata/blob/ee12a8b314779362afe22e9aa5f6aa431ad5810a/fermata/src/auto/java/me/aap/fermata/auto/Xposed.java#L58

It should be:
```java
findAndHookMethod(PackageManager.class, "getInstallerPackageName", String.class, new XC_MethodHook() {
```

Can you build the app? If no, I can build it for you.

**Comment by roger2k13324** [comment]
Great find! I do not have a build environment set up yet, so if it is not a big issue for you, I would happily accept your build. Thanks!

**Reply by AndreyPavlenko** [reply]
https://app.box.com/s/ns86n7vc3ssaltli0mmibrydnyvbr6bs

**Reply by roger2k13324** [reply]
Sadly, it seems like hooking abstract methods is not possible:

```
02-14 14:59:07.363 I/Xposed  ( 5773): Hooking package com.google.android.projection.gearhead
02-14 14:59:07.365 E/Xposed  ( 5773): java.lang.IllegalArgumentException: Cannot hook abstract methods: public abstract java.lang.String android.content.pm.PackageManager.getInstallerPackageName(java.lang.String)
```

```
02-14 14:59:06.407 I/Xposed  ( 5563): Hooking package com.android.printspooler
02-14 14:59:06.411 E/Xposed  ( 5556): java.lang.IllegalArgumentException: Cannot hook abstract methods: public abstract java.lang.String android.content.pm.PackageManager.getInstallerPackageName(java.lang.String)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.XposedBridge.hookMethod(XposedBridge.java:210)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.XposedHelpers.findAndHookMethod(XposedHelpers.java:187)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at me.aap.fermata.auto.Xposed.handleLoadPackage(SourceFile:75)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.IXposedHookLoadPackage$Wrapper.handleLoadPackage(IXposedHookLoadPackage.java:34)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.callbacks.XC_LoadPackage.call(XC_LoadPackage.java:61)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.callbacks.XCallback.callAll(XCallback.java:106)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.XposedInit$2.beforeHookedMethod(XposedInit.java:134)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.XposedBridge.handleHookedMethod(XposedBridge.java:341)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.app.ActivityThread.handleBindApplication(<Xposed>)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.app.ActivityThread.-wrap1(Unknown Source:0)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1764)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.os.Handler.dispatchMessage(Handler.java:105)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.os.Looper.loop(Looper.java:164)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at android.app.ActivityThread.main(ActivityThread.java:6944)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at java.lang.reflect.Method.invoke(Native Method)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
02-14 14:59:06.411 E/Xposed  ( 5556): 	at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:108)
```

**Reply by roger2k13324** [reply]
The implementation class is maybe `android.content.pm.IPackageManager.Stub.Proxy`.
https://stackoverflow.com/questions/52126563/how-to-hook-the-abstract-method-hassystemfeature

**Reply by AndreyPavlenko** [reply]
Changed the class name - https://app.box.com/s/ns86n7vc3ssaltli0mmibrydnyvbr6bs

**Reply by roger2k13324** [reply]
The exception is gone and the Fermata apps are visible under 'Customize launcher' even if the phone is disconnected from DHU! Will try in the car in a few minutes. Thanks!

**Reply by roger2k13324** [reply]
The issue is fixed, the apps are visible when connected to the car HU. Thank you!
Regarding the other apps "Fermata" and "Fermata mirror", do you have any idea why they do not work for me? Maybe the Android version is too low? It seems like the apps are crashing on startup. I will have a look at the logcat output.
May I ask whether the normal "Fermata" app is also using screen mirroring behind the scenes? Although the "Fermata Mirror FS" app is fully operational, fast moving scenes have a bit of stutter, which could be caused by the weak performance of my old phone (video decoding & encoding for mirror).
Thanks!

**Reply by AndreyPavlenko** [reply]
You can open the log in Settings > Other.
No, Fermata Auto is not a mirroring app.

**Reply by roger2k13324** [reply]
It seems Fermata does not produce any logs, I assume the application never gets to the point where it is running normally.
I have attached the logcat output of the whole process lifecycle filtered by PID to reduce size.
There are a few exceptions which could be connected to the issue:

`java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.car.CarDisplayInfoManager" on path: DexPathList[[zip file "/data/user/0/me.aap.fermata.auto.dear.google.why/app_car_sdk_impl/sdk_impl.jar"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]`

`java.lang.IncompatibleClassChangeError: The method 'java.lang.Class java.lang.Object.getClass()' was expected to be of type interface but instead was found to be of type virtual (declaration of 'com.google.android.gms.a.d' appears in /data/data/me.aap.fermata.auto.dear.google.why/app_car_sdk_impl/sdk_impl.jar)`

`java.lang.SecurityException: projection client manager does not have permission:android.permission.CAPTURE_VIDEO_OUTPUT pid:0 uid:10197`

For the last one, I tried to insert `android.permission.CAPTURE_VIDEO_OUTPUT` into the manifest and grant with `pm grant` but granting fails with an error saying this permission is not changeable for normal packages.

Do you have any idea how I could further investigate?

Thanks!

[logcat.txt](https://github.com/AndreyPavlenko/Fermata/files/14293938/logcat.txt)

**Reply by AndreyPavlenko** [reply]
Never seen this error before. Is it only happening in Android Auto? Can you start the app on your phone?

**Reply by roger2k13324** [reply]
Yes, I can start the app on the phone when not connected to the HU and it works perfectly. When connected to the HU, I can start the app on the phone by tapping on its icon on the phone screen and it works (on the phone display) but when I try to start it by tapping on its icon on the HU, the HU displays a black surface and AA hangs. However, with "Fermata Mirror FS" the "Fermata" app content is visible and everything is functional. Downside is that the phone has to be unlocked all the time plus the additional HW usage because of mirroring.
I can also record a video to show the behavior when needed.

**Reply by AndreyPavlenko** [reply]
Seems sdk is not compatible with your system, for some reason. Could you try a different firmware?
